File: mvn-advent-of-code/src/test/java/advent/twenty_fifteen/Day08MatchsticksTest.java
package advent.twenty_fifteen;

import advent.common.DailyProblem;
import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class Day08MatchsticksTest {
    @Test
    public void testEncode() {
        assertEquals("\"\\\"\\\"\"", Day08Matchsticks.encode("\"\""));
        assertEquals("\"\\\"abc\\\"\"", Day08Matchsticks.encode("\"abc\""));
        assertEquals("\"\\\"aaa\\\\\\\"aaa\\\"\"", Day08Matchsticks.encode("\"aaa\\\"aaa\""));
        assertEquals("\"\\\"\\\\x27\\\"\"", Day08Matchsticks.encode("\"\\x27\""));
    }

    @Test
    public void testDecode() {
        assertEquals("", Day08Matchsticks.decode("\"\""));
        assertEquals("abc", Day08Matchsticks.decode("\"abc\""));
        assertEquals("aaa\"aaa", Day08Matchsticks.decode("\"aaa\\\"aaa\""));
        assertEquals("'", Day08Matchsticks.decode("\"\\x27\""));
    }

    @Test
    public void checkReal() {
        DailyProblem<Integer, Integer> dp = new Day08Matchsticks("/twenty_fifteen/Day08-Matchsticks-input.txt");

        assertEquals(1350, dp.getPart1Answer().intValue());
        assertEquals(2085, dp.getPart2Answer().intValue());
    }

    @Test
    public void checkExample1() {
        DailyProblem<Integer, Integer> dp = new Day08Matchsticks("/twenty_fifteen/Day08-Matchsticks-example#1.txt");

        assertEquals(12, dp.getPart1Answer().intValue());
        assertEquals(19, dp.getPart2Answer().intValue());
    }
}

File: scala-advent-of-code/src/test/scala/advent/twenty_fifteen/Day08MatchsticksTest.scala
package advent.twenty_fifteen

import org.scalatest.{FlatSpec, Matchers}

class Day08MatchsticksTest extends FlatSpec with Matchers {
  "2015 Day 8 - Input" should "calculate answers" in {
    val dp = new Day08Matchsticks("/twenty_fifteen/Day08-Matchsticks-input.txt")

    dp.part1Answer should be(1350)
    dp.part2Answer should be(2085)
  }

  "2015 Day 8 - Example #1" should "calculate answers" in {
    val dp = new Day08Matchsticks("/twenty_fifteen/Day08-Matchsticks-example#1.txt")

    dp.part1Answer should be(12)
    dp.part2Answer should be(19)
  }

  "2015 Day 8 - Encode" should "calculate answers" in {
    val dp = new Day08Matchsticks("/twenty_fifteen/Day08-Matchsticks-example#1.txt")

    dp.encode("\"abc\"", "") should be("\"\\\"abc\\\"\"")
    dp.encode("\"aaa\\\"aaa\"", "") should be("\"\\\"aaa\\\\\\\"aaa\\\"\"")
    dp.encode("\"\\x27\"", "") should be("\"\\\"\\\\x27\\\"\"")
  }

  "2015 Day 8 - Decode" should "calculate answers" in {
    val dp = new Day08Matchsticks("/twenty_fifteen/Day08-Matchsticks-example#1.txt")

    dp.decode("\"abc\"", "") should be("abc")
    dp.decode("\"aaa\\\"aaa\"", "") should be("aaa\"aaa")
    dp.decode("\"\\x27\"", "") should be("'")
  }
}
